ush/ioda/bufr2ioda/bufr2ioda_adpupa_prepbufr.py

```diff
@@ -0,0 +1,444 @@
+# (C) Copyright 2023 NOAA/NWS/NCEP/EMC
+#
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+
+import sys
+import argparse
+import numpy as np
+import numpy.ma as ma
+import calendar
+import json
+import time
+import math
+import datetime
+import os
+from datetime import datetime
+from pyioda import ioda_obs_space as ioda_ospace
+from wxflow import Logger
+from pyiodaconv import bufr
+from collections import namedtuple
+import warnings
+  
+# suppress warnings
+warnings.filterwarnings('ignore')
+
+#|------------------------------------------------------------------------------|
+#|            gdas prepbufr file                                                |                    
+#|------------------------------------------------------------------------------|
+#| MNEMONIC | NUMBER | DESCRIPTION                                              |
+#|----------|--------|----------------------------------------------------------|
+#|          |        |                                                          |
+#| ADPUPA   | A48102 | UPPER-AIR (RAOB, PIBAL, RECCO, DROPS) REPORTS            |
+#|------------------------------------------------------------------------------|
+
+
+def Compute_dateTime(cycleTimeSinceEpoch, hrdr):
+
+    hrdr = np.int64(hrdr*3600)
+    hrdr = hrdr+ cycleTimeSinceEpoch
+
+    return hrdr
+
+
+def bufr_to_ioda(config, logger):
+
+    subsets = config["subsets"]
+    logger.debug(f"Checking subsets = {subsets}")
+
+    # Get parameters from configuration
+    subsets 		= config["subsets"]
+    data_format 	= config["data_format"]
+    data_type 		= config["data_type"]
+    data_description 	= config["data_description"]
+    data_provider 	= config["data_provider"]
+    cycle_type 		= config["cycle_type"]
+    dump_dir 		= config["dump_directory"]
+    ioda_dir 		= config["ioda_directory"]
+    cycle 		= config["cycle_datetime"]
+
+    # Get derived parameters
+    yyyymmdd 		= cycle[0:8]
+    hh 			= cycle[8:10]
+    reference_time 	= datetime.strptime(cycle, "%Y%m%d%H")
+    reference_time 	= reference_time.strftime("%Y-%m-%dT%H:%M:%SZ")
+
+    # General informaton
+    converter 			= 'prepBUFR to IODA Converter'
+    platform_description 	= 'ADPUPA prepbufr'
+
+    logger.info(f"reference_time = {reference_time}")
+
+    bufrfile 			= f"{cycle_type}.t{hh}z.{data_type}"
+    DATA_PATH = os.path.join(dump_dir, f"{cycle_type}.{yyyymmdd}", str(hh), 'atmos', bufrfile)
+    #gdas.t18z.prepbufr  gdas.t18z.satwnd.tm00.bufr_d
+    #bufrfile = f"{cycle_type}.t{hh}z.{data_type}.tm00.{data_format}"
+
+    # ============================================
+    # Make the QuerySet for all the data we want
+    # ============================================
+    start_time = time.time()
+
+    logger.info('Making QuerySet')
+    q = bufr.QuerySet(subsets)
+
+    # MetaData
+    q.add('latitude', 					'*/PRSLEVEL/DRFTINFO/YDR')
```

Comment:
is there a reason why these are "tab'd" so far to the right?

Reply:
@CoryMartin-NOAA there is no specific reason, just for clarity.

Reply:
@CoryMartin-NOAA finally removed all the "tab'd" as those were one of the reasons for the failure of unit test(s).